File: src/agent/memory.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { embedText, l2Normalize } from '@/agent/embeddings';
import { supabase } from '@/lib/supabase';
import { getOperatingPicture, ragSearch } from '@/services/memory';
import type { OperatingPicture, RagResult } from '@/services/memory';
import type { RoutedIntent } from '@/agent/types';

const SQLITE_DB_NAME = 'riflett-memory.db';
const STORAGE_KEY = 'riflett_memory_store';
const MAX_CACHED_ROWS = 512;
const REMOTE_BRIEF_LIMIT = 9;

const resolveUserId = async (): Promise<string | null> => {
  const {
    data: { user },
    error,
  } = await supabase.auth.getUser();
  if (error) {
    console.warn('[memory] resolveUserId failed', error);
    return null;
  }
  return user?.id ?? null;
};

type SQLiteResultSet = {
  rows: {
    length: number;
    item: (index: number) => any;
    _array: any[];
  };
  rowsAffected?: number | undefined;
  insertId?: number | undefined;
};

type LegacySQLiteTransaction = {
  executeSql: (
    sql: string,
    params: any[],
    success?: (tx: LegacySQLiteTransaction, result: SQLiteResultSet) => void,
    error?: (tx: LegacySQLiteTransaction, error: Error) => boolean
  ) => void;
};

type LegacySQLiteDatabase = {
  transaction: (
    callback: (tx: LegacySQLiteTransaction) => void,
    error?: (error: Error) => void,
    success?: () => void
  ) => void;
};

type ModernSQLiteExecuteResult<T = any> = {
  getAllAsync: () => Promise<T[]>;
  getFirstAsync: () => Promise<T | null>;
  resetAsync: () => Promise<void>;
  readonly changes: number;
  readonly lastInsertRowId: number;
};

type ModernSQLiteStatement = {
  executeAsync: (...params: any[]) => Promise<ModernSQLiteExecuteResult>;
  finalizeAsync: () => Promise<void>;
};

type ModernSQLiteDatabase = {
  prepareAsync: (sql: string) => Promise<ModernSQLiteStatement>;
};

type ExpoSqliteModule = Partial<
  {
    openDatabase: (name: string) => LegacySQLiteDatabase;
    openDatabaseAsync: (name: string) => Promise<ModernSQLiteDatabase>;
    openDatabaseSync: (name: string) => ModernSQLiteDatabase;
  }
>;

type SQLiteDatabase = LegacySQLiteDatabase | ModernSQLiteDatabase;

let sqliteDb: SQLiteDatabase | null = null;
let sqliteFlavor: 'legacy' | 'modern' | null = null;
let sqliteReady = false;
let initialized: Promise<void> | null = null;

const memoryCache = new Map<string, MemoryRow>();

interface MemoryRow {
  id: string;
  kind: MemoryKind;
  text: string;
  ts: number;
  embedding: number[];
}

const mapKindsToRag = (kinds: string[]): RagKind[] => {
  const set = new Set(kinds.map((kind) => kind.toLowerCase()))
  const mapped: RagKind[] = []
  if (set.has('entry') || set.has('journal') || set.has('pref')) {
    mapped.push('entry')
  }
  if (set.has('goal')) {
    mapped.push('goal')
  }
  if (set.has('schedule') || set.has('event')) {
    mapped.push('schedule')
  }
  return mapped
}

const inferScopeFromIntent = (intent: RoutedIntent | null | undefined): RagScopeInput => {
  if (!intent) return 'all'
  const label = intent.label?.toLowerCase() ?? ''
  const scopes: RagKind[] = []
  if (label.includes('goal')) {
    scopes.push('goal')
  }
  if (label.includes('schedule') || label.includes('calendar')) {
    scopes.push('schedule')
  }
  if (scopes.length === 0 || label.includes('journal') || label.includes('reflect')) {
    scopes.push('entry')
  }
  return scopes.length ? scopes : 'all'
}

export type MemoryKind = 'entry' | 'goal' | 'event' | 'pref' | 'schedule';
type RagKind = 'entry' | 'goal' | 'schedule';
type RagScopeInput = RagKind | RagKind[] | 'all';

export interface MemoryRecord extends MemoryRow {
  score: number;
}

interface SearchOptions {
  query: string;
  kinds: string[];
  topK: number;
}

interface UpsertOptions {
  id: string;
  kind: MemoryKind;
  text: string;
  ts?: number;
  embedding?: number[];
}

const executeSql = async (sql: string, params: any[] = []): Promise<SQLiteResultSet | null> => {
  if (!sqliteDb) return null;

  if (sqliteFlavor === 'legacy' && sqliteDb && isLegacyDatabase(sqliteDb)) {
    const legacyDb = sqliteDb;
    return new Promise((resolve, reject) => {
      legacyDb.transaction(
        (tx: LegacySQLiteTransaction) => {
          tx.executeSql(
            sql,
            params,
            (_tx: LegacySQLiteTransaction, result: SQLiteResultSet) => {
              resolve(result);
            },
            (_tx: LegacySQLiteTransaction, error: Error) => {
              reject(error);
              return false;
            }
          );
        },
        (error: Error) => reject(error)
      );
    });
  }

  if (sqliteFlavor === 'modern' && sqliteDb && isModernDatabase(sqliteDb)) {
    const modernDb = sqliteDb;
    let statement: ModernSQLiteStatement | null = null;
    try {
      statement = await modernDb.prepareAsync(sql);
      const iterator = await statement.executeAsync(params);
      const rowsArray = await iterator.getAllAsync();
      const result: SQLiteResultSet = {
        rows: {
          length: rowsArray.length,
          item: (index: number) => rowsArray[index],
          _array: rowsArray,
        },
      };
      if (typeof iterator.changes === 'number') {
        result.rowsAffected = iterator.changes;
      }
      if (typeof iterator.lastInsertRowId === 'number' && Number.isFinite(iterator.lastInsertRowId)) {
        result.insertId = iterator.lastInsertRowId;
      }
      return result;
    } finally {
      if (statement) {
        try {
          await statement.finalizeAsync();
        } catch (finalizeError) {
          console.warn('[memory] Failed to finalize SQLite statement', finalizeError);
        }
      }
    }
  }

  return null;
};

const loadSqliteModule = (): ExpoSqliteModule | null => {
  try {
    // Use CommonJS require to avoid Metro async shim.
    const baseModule = require('expo-sqlite') as ExpoSqliteModule & {
      default?: ExpoSqliteModule;
      legacy?: ExpoSqliteModule;
    };

    const candidates = [baseModule, baseModule?.default, baseModule?.legacy].filter(
      (candidate): candidate is ExpoSqliteModule =>
        !!candidate && typeof candidate === 'object'
    );

    for (const candidate of candidates) {
      if (
        typeof candidate.openDatabaseAsync === 'function' ||
        typeof candidate.openDatabaseSync === 'function' ||
        typeof candidate.openDatabase === 'function'
      ) {
        return candidate;
      }
    }

    console.warn('[memory] expo-sqlite module loaded without an openDatabase entrypoint');
    return null;
  } catch (error) {
    console.warn('[memory] expo-sqlite unavailable', error);
    return null;
  }
};

const isLegacyDatabase = (db: SQLiteDatabase): db is LegacySQLiteDatabase =>
  typeof (db as LegacySQLiteDatabase).transaction === 'function';

const isModernDatabase = (db: SQLiteDatabase): db is ModernSQLiteDatabase =>
  typeof (db as ModernSQLiteDatabase).prepareAsync === 'function';

const tryInitSqlite = async (): Promise<boolean> => {
  if (sqliteReady) return true;

  try {
    const sqliteModule = loadSqliteModule();
    if (!sqliteModule) {
      sqliteDb = null;
      sqliteFlavor = null;
      sqliteReady = false;
      return false;
    }
    if (typeof sqliteModule.openDatabaseAsync === 'function') {
      sqliteDb = await sqliteModule.openDatabaseAsync(SQLITE_DB_NAME);
    } else if (typeof sqliteModule.openDatabaseSync === 'function') {
      sqliteDb = sqliteModule.openDatabaseSync(SQLITE_DB_NAME);
    } else if (typeof sqliteModule.openDatabase === 'function') {
      sqliteDb = sqliteModule.openDatabase(SQLITE_DB_NAME);
    } else {
      throw new Error('expo-sqlite module does not expose an openDatabase function');
    }

    if (sqliteDb && isModernDatabase(sqliteDb)) {
      sqliteFlavor = 'modern';
    } else if (sqliteDb && isLegacyDatabase(sqliteDb)) {
      sqliteFlavor = 'legacy';
    } else {
      throw new Error('expo-sqlite returned an unsupported database instance');
    }

    await executeSql(
      'CREATE TABLE IF NOT EXISTS memory (id TEXT PRIMARY KEY, kind TEXT, text TEXT, ts INTEGER, embedding TEXT)'
    );
    await executeSql(
      'CREATE INDEX IF NOT EXISTS idx_memory_kind_ts ON memory(kind, ts DESC)'
    );

    const result = await executeSql('SELECT id, kind, text, ts, embedding FROM memory');
    if (result && Array.isArray(result.rows?._array)) {
      result.rows._array.forEach((row: any) => {
        const parsed: MemoryRow = {
          id: row.id,
          kind: row.kind,
          text: row.text,
          ts: row.ts,
          embedding: parseEmbedding(row.embedding),
        };
        memoryCache.set(parsed.id, parsed);
      });
    }

    sqliteReady = true;
    return true;
  } catch (error) {
    console.warn('[memory] SQLite init failed, using AsyncStorage fallback', error);
    sqliteDb = null;
    sqliteFlavor = null;
    sqliteReady = false;
    return false;
  }
};

const parseEmbedding = (raw: unknown): number[] => {
  if (Array.isArray(raw)) {
    return l2Normalize(raw.map((v) => Number(v)));
  }
  if (typeof raw === 'string') {
    try {
      const parsed = JSON.parse(raw) as number[];
      return l2Normalize(parsed.map((v) => Number(v)));
    } catch (error) {
      console.warn('[memory] Failed to parse embedding JSON', error);
    }
  }
  return [];
};

const serializeEmbedding = (embedding: number[]): string => JSON.stringify(embedding);

const ensureInitialized = async (): Promise<void> => {
  if (!initialized) {
    initialized = (async () => {
      const sqliteAvailable = await tryInitSqlite();
      if (!sqliteAvailable) {
        await hydrateFromStorage();
      }
    })();
  }
  return initialized;
};

const hydrateFromStorage = async (): Promise<void> => {
  try {
    const raw = await AsyncStorage.getItem(STORAGE_KEY);
    if (!raw) return;
    const parsed = JSON.parse(raw) as MemoryRow[];
    parsed.forEach((row) => {
      memoryCache.set(row.id, {
        ...row,
        embedding: Array.isArray(row.embedding)
          ? l2Normalize(row.embedding.map((v) => Number(v)))
          : [],
      });
    });
  } catch (error) {
    console.warn('[memory] hydrate fallback failed', error);
  }
};

const persistFallback = async (): Promise<void> => {
  const subset = Array.from(memoryCache.values())
    .sort((a, b) => b.ts - a.ts)
    .slice(0, MAX_CACHED_ROWS)
    .map((row) => ({
      ...row,
      embedding: row.embedding,
    }));
  try {
    await AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(subset));
  } catch (error) {
    console.warn('[memory] persist fallback failed', error);
  }
};

const cosineScores = (query: number[], rows: MemoryRow[]): MemoryRecord[] => {
  const normalizedQuery = l2Normalize(query);
  const scores = rows.map((row) => {
    const embedding = row.embedding;
    const length = Math.min(normalizedQuery.length, embedding.length);
    let dot = 0;
    for (let i = 0; i < length; i += 1) {
      dot += normalizedQuery[i]! * embedding[i]!;
    }
    return {
      ...row,
      score: dot,
    };
  });

  return scores.sort((a, b) => b.score - a.score);
};

export const Memory = {
  async searchTopN(options: SearchOptions): Promise<MemoryRecord[]> {
    await ensureInitialized();
    const topK = Math.max(1, Math.min(options.topK || 5, 20));
    const userId = await resolveUserId();
    if (userId) {
      try {
        const scopeKinds = mapKindsToRag(options.kinds);
        const ragScope = scopeKinds.length ? scopeKinds : 'all';
        const ragResults = await ragSearch(
          userId,
          options.query,
          ragScope,
          { limit: topK }
        );
        if (ragResults.length) {
          const nowTs = Date.now();
          const remoteRecords: MemoryRecord[] = [];
          for (const [index, result] of ragResults.entries()) {
            const text = result.snippet ?? '';
            let embedding: number[] = [];
            try {
              embedding = await embedText(text);
            } catch (error) {
              console.warn('[memory] embedding cache failure', error);
            }
            const record: MemoryRecord = {
              id: `${result.kind}:${result.id}`,
              kind: (result.kind as MemoryKind) ?? 'entry',
              text,
              ts: nowTs - index,
              embedding,
              score: result.score,
            };
            remoteRecords.push(record);
            try {
              await Memory.upsert({
                id: record.id,
                kind: record.kind,
                text: record.text,
                ts: record.ts,
                embedding: record.embedding,
              });
            } catch (error) {
              console.warn('[memory] remote cache upsert failed', error);
            }
          }
          return remoteRecords.slice(0, topK);
        }
      } catch (error) {
        console.warn('[memory] remote search failed, falling back', error);
      }
    }

    const kinds = options.kinds.map((kind) => kind.toLowerCase());
    const queryVector = await embedText(options.query);

    if (sqliteReady && sqliteDb) {
      const placeholders = kinds.length
        ? kinds.map(() => '?').join(',')
        : '?';
      const params = kinds.length ? kinds : ['entry'];
      const result = await executeSql(
        `SELECT id, kind, text, ts, embedding FROM memory WHERE kind IN (${placeholders}) ORDER BY ts DESC LIMIT 512`,
        params
      );

      const rows: MemoryRow[] = result && result.rows
        ? result.rows._array.map((row: any) => ({
            id: row.id,
            kind: row.kind,
            text: row.text,
            ts: row.ts,
            embedding: parseEmbedding(row.embedding),
          }))
        : [];

      return cosineScores(queryVector, rows).slice(0, topK);
    }

    const rows = Array.from(memoryCache.values()).filter((row) =>
      kinds.length ? kinds.includes(row.kind.toLowerCase()) : true
    );

    return cosineScores(queryVector, rows).slice(0, topK);
  },

  async getBrief(
    uid: string | null,
    intent: RoutedIntent,
    query: string,
    options: { limit?: number } = {}
  ): Promise<{
    operatingPicture: OperatingPicture;
    rag: RagResult[];
    memoryRecords: MemoryRecord[];
  }> {
    await ensureInitialized();
    const userId = uid ?? (await resolveUserId());
    if (!userId) {
      throw new Error('User not authenticated');
    }

    const limit = Math.max(3, Math.min(options.limit ?? REMOTE_BRIEF_LIMIT, REMOTE_BRIEF_LIMIT));
    const [operatingPicture, ragResults] = await Promise.all([
      getOperatingPicture(userId),
      ragSearch(userId, query, inferScopeFromIntent(intent), { limit }),
    ]);

    const memoryRecords: MemoryRecord[] = [];
    const nowTs = Date.now();

    for (const [index, result] of ragResults.entries()) {
      const text = result.snippet ?? '';
      let embedding: number[] = [];
      try {
        embedding = await embedText(text);
      } catch (error) {
        console.warn('[memory] brief embedding failed', error);
      }
      const record: MemoryRecord = {
        id: `${result.kind}:${result.id}`,
        kind: (result.kind as MemoryKind) ?? 'entry',
        text,
        ts: nowTs - index,
        embedding,
        score: result.score,
      };
      memoryRecords.push(record);
      try {
        await Memory.upsert({
          id: record.id,
          kind: record.kind,
          text: record.text,
          ts: record.ts,
          embedding: record.embedding,
        });
      } catch (error) {
        console.warn('[memory] brief cache upsert failed', error);
      }
    }

    return {
      operatingPicture,
      rag: ragResults,
      memoryRecords,
    };
  },

  async upsert(options: UpsertOptions): Promise<void> {
    await ensureInitialized();

    const ts = options.ts ?? Date.now();
    const embedding = options.embedding ?? (await embedText(options.text));
    const normalized = l2Normalize(embedding);
    const row: MemoryRow = {
      id: options.id,
      kind: options.kind,
      text: options.text,
      ts,
      embedding: normalized,
    };

    memoryCache.set(row.id, row);

    if (sqliteReady && sqliteDb) {
      await executeSql(
        'REPLACE INTO memory (id, kind, text, ts, embedding) VALUES (?, ?, ?, ?, ?)',
        [row.id, row.kind, row.text, row.ts, serializeEmbedding(row.embedding)]
      );
    } else {
      await persistFallback();
    }
  },

  async remove(id: string): Promise<void> {
    await ensureInitialized();
    memoryCache.delete(id);
    if (sqliteReady && sqliteDb) {
      await executeSql('DELETE FROM memory WHERE id = ?', [id]);
    } else {
      await persistFallback();
    }
  },

  async warmup(): Promise<void> {
    await ensureInitialized();
    // Trigger a trivial embedding to warm up the model / thread
    await embedText('hi');
  },
};

export type MemorySearchResult = Awaited<ReturnType<typeof Memory.searchTopN>>;

File: services/python-ai/app/agent/memory.py
import time
from dataclasses import dataclass
from typing import Dict, List, Literal, Optional

from .embeddings import embed_text, l2_normalize
from .types import RoutedIntent

MemoryKind = Literal["entry", "goal", "event", "pref", "schedule"]
RagKind = Literal["entry", "goal", "schedule"]
RagScopeInput = RagKind | List[RagKind] | Literal["all"]

MAX_CACHED_ROWS = 512
REMOTE_BRIEF_LIMIT = 9

_memory_cache: Dict[str, "MemoryRow"] = {}
_rag_results: List[Dict[str, object]] = []
_operating_picture: Optional[Dict[str, object]] = None
_user_id: Optional[str] = None


@dataclass
class MemoryRow:
    id: str
    kind: MemoryKind
    text: str
    ts: int
    embedding: List[float]


@dataclass
class MemoryRecord(MemoryRow):
    score: float


def set_authenticated_user(user_id: Optional[str]) -> None:
    global _user_id
    _user_id = user_id


def seed_rag_results(results: List[Dict[str, object]]) -> None:
    global _rag_results
    _rag_results = results


def seed_operating_picture(picture: Optional[Dict[str, object]]) -> None:
    global _operating_picture
    _operating_picture = picture


async def resolve_user_id() -> Optional[str]:
    return _user_id


def _map_kinds_to_rag(kinds: List[str]) -> List[RagKind]:
    normalized = {kind.lower() for kind in kinds}
    mapped: List[RagKind] = []
    if {"entry", "journal", "pref"} & normalized:
        mapped.append("entry")
    if "goal" in normalized:
        mapped.append("goal")
    if {"schedule", "event"} & normalized:
        mapped.append("schedule")
    return mapped


def _infer_scope_from_intent(intent: Optional[RoutedIntent]) -> RagScopeInput:
    if not intent:
        return "all"
    label = (intent.label or "").lower()
    scopes: List[RagKind] = []
    if "goal" in label:
        scopes.append("goal")
    if "schedule" in label or "calendar" in label:
        scopes.append("schedule")
    if not scopes or "journal" in label or "reflect" in label:
        scopes.append("entry")
    return scopes or "all"


def _cosine_scores(query: List[float], rows: List[MemoryRow]) -> List[MemoryRecord]:
    normalized_query = l2_normalize(query)
    scores: List[MemoryRecord] = []
    for row in rows:
        embedding = row.embedding
        length = min(len(normalized_query), len(embedding))
        dot = sum(normalized_query[i] * embedding[i] for i in range(length))
        scores.append(MemoryRecord(**row.__dict__, score=dot))
    return sorted(scores, key=lambda record: record.score, reverse=True)


async def _rag_search(user_id: str, query: str, scope: RagScopeInput = "all", options: Optional[Dict[str, object]] = None) -> List[Dict[str, object]]:
    trimmed = query.strip()
    if not trimmed:
        return []
    limit = options.get("limit") if options else None
    results = list(_rag_results)
    if isinstance(limit, int):
        results = results[:limit]
    return [
        {
            "id": str(item.get("id", "")),
            "kind": item.get("kind", "entry"),
            "score": float(item.get("score", 0)),
            "title": item.get("title"),
            "snippet": str(item.get("snippet", "")),
            "metadata": item.get("metadata") or {},
        }
        for item in results
    ]


async def _get_operating_picture(user_id: str) -> Optional[Dict[str, object]]:
    return _operating_picture


def _default_operating_picture() -> Dict[str, object]:
    return {
        "why_model": None,
        "top_goals": [],
        "hot_entries": [],
        "next_72h": [],
        "cadence_profile": {
            "cadence": "none",
            "session_length_minutes": 25,
            "last_message_at": None,
            "missed_day_count": 0,
            "current_streak": 0,
            "timezone": "UTC",
        },
        "risk_flags": [],
    }


class Memory:
    @staticmethod
    async def search_top_n(options: Dict[str, object]) -> List[MemoryRecord]:
        query = str(options.get("query", "")).strip()
        if not query:
            return []
        top_k = max(1, min(int(options.get("topK", 5)), 20))
        kinds = [kind.lower() for kind in options.get("kinds", [])] if isinstance(options.get("kinds"), list) else []
        user_id = await resolve_user_id()

        if user_id:
            try:
                scope_kinds = _map_kinds_to_rag(kinds)
                rag_scope = scope_kinds if scope_kinds else "all"
                rag_results = await _rag_search(user_id, query, rag_scope, {"limit": top_k})
                if rag_results:
                    now_ts = int(time.time() * 1000)
                    remote_records: List[MemoryRecord] = []
                    for index, result in enumerate(rag_results):
                        text = result.get("snippet") or ""
                        embedding = await embed_text(text)
                        record = MemoryRecord(
                            id=f"{result.get('kind')}:{result.get('id')}",
                            kind=result.get("kind", "entry"),  # type: ignore[arg-type]
                            text=text,
                            ts=now_ts - index,
                            embedding=embedding,
                            score=float(result.get("score", 0)),
                        )
                        remote_records.append(record)
                        try:
                            await Memory.upsert(
                                {"id": record.id, "kind": record.kind, "text": record.text, "ts": record.ts, "embedding": record.embedding}
                            )
                        except Exception:
                            pass
                    return remote_records[:top_k]
            except Exception:
                pass

        query_vector = await embed_text(query)
        rows = [row for row in _memory_cache.values() if not kinds or row.kind.lower() in kinds]
        return _cosine_scores(query_vector, rows)[:top_k]

    @staticmethod
    async def get_brief(uid: Optional[str], intent: RoutedIntent, query: str, options: Optional[Dict[str, object]] = None):
        user_id = uid or await resolve_user_id()
        if not user_id:
            raise ValueError("User not authenticated")

        limit = max(3, min(int(options.get("limit", REMOTE_BRIEF_LIMIT)) if options else REMOTE_BRIEF_LIMIT, REMOTE_BRIEF_LIMIT))

        operating_picture = options.get("cachedOperatingPicture") if options else None
        if operating_picture is None:
            try:
                operating_picture = await _get_operating_picture(user_id)
            except Exception:
                operating_picture = None

        if operating_picture is None:
            operating_picture = _default_operating_picture()

        rag_results = await _rag_search(user_id, query, _infer_scope_from_intent(intent), {"limit": limit})
        memory_records: List[MemoryRecord] = []
        now_ts = int(time.time() * 1000)

        for index, result in enumerate(rag_results):
            text = result.get("snippet") or ""
            try:
                embedding = await embed_text(text)
            except Exception:
                embedding = []
            record = MemoryRecord(
                id=f"{result.get('kind')}:{result.get('id')}",
                kind=result.get("kind", "entry"),  # type: ignore[arg-type]
                text=text,
                ts=now_ts - index,
                embedding=embedding,
                score=float(result.get("score", 0)),
            )
            memory_records.append(record)
            try:
                await Memory.upsert(
                    {"id": record.id, "kind": record.kind, "text": record.text, "ts": record.ts, "embedding": record.embedding}
                )
            except Exception:
                pass

        return {"operatingPicture": operating_picture, "rag": rag_results, "memoryRecords": memory_records}

    @staticmethod
    async def upsert(options: Dict[str, object]) -> None:
        ts = int(options.get("ts", int(time.time() * 1000)))
        embedding = options.get("embedding") or await embed_text(str(options.get("text", "")))
        normalized = l2_normalize(embedding)
        row = MemoryRow(
            id=str(options["id"]),
            kind=options["kind"],  # type: ignore[arg-type]
            text=str(options.get("text", "")),
            ts=ts,
            embedding=normalized,
        )
        _memory_cache[row.id] = row
        if len(_memory_cache) > MAX_CACHED_ROWS:
            oldest = sorted(_memory_cache.values(), key=lambda value: value.ts, reverse=True)[:MAX_CACHED_ROWS]
            _memory_cache.clear()
            _memory_cache.update({row.id: row for row in oldest})

    @staticmethod
    async def remove(row_id: str) -> None:
        _memory_cache.pop(row_id, None)

    @staticmethod
    async def warmup() -> None:
        await embed_text("hi")


MemorySearchResult = List[MemoryRecord]
